File: app/src/main/java/net/squanchy/injection/ApplicationComponent.java
package net.squanchy.injection;

import android.app.Application;

import net.squanchy.analytics.Analytics;
import net.squanchy.analytics.AnalyticsModule;
import net.squanchy.remoteconfig.RemoteConfig;
import net.squanchy.remoteconfig.RemoteConfigModule;
import net.squanchy.service.firebase.FirebaseAuthService;
import net.squanchy.service.firebase.FirebaseDbService;
import net.squanchy.service.firebase.injection.FirebaseModule;
import net.squanchy.service.repository.EventRepository;
import net.squanchy.service.repository.SpeakerRepository;
import net.squanchy.service.repository.VenueRepository;
import net.squanchy.service.repository.injection.RepositoryModule;
import net.squanchy.support.injection.ChecksumModule;
import net.squanchy.support.injection.CurrentTimeModule;

import dagger.Component;

@ApplicationLifecycle
@Component(modules = {
        ApplicationContextModule.class,
        FirebaseModule.class,
        ChecksumModule.class,
        RepositoryModule.class,
        AnalyticsModule.class,
        RemoteConfigModule.class,
        CurrentTimeModule.class})
public interface ApplicationComponent {

    FirebaseDbService firebaseDbService();

    FirebaseAuthService firebaseAuthService();

    EventRepository eventRepository();

    SpeakerRepository speakerRepository();

    VenueRepository venueRepository();

    Analytics analytics();

    RemoteConfig remoteConfig();

    Application application();

    class Factory {

        private Factory() {
            // non-instantiable
        }

        public static ApplicationComponent create(Application application) {
            return DaggerApplicationComponent.builder()
                    .firebaseModule(new FirebaseModule())
                    .repositoryModule(new RepositoryModule())
                    .checksumModule(new ChecksumModule())
                    .applicationContextModule(new ApplicationContextModule(application))
                    .analyticsModule(new AnalyticsModule(application))
                    .remoteConfigModule(new RemoteConfigModule())
                    .build();
        }
    }
}

File: app/src/main/java/net/squanchy/injection/ApplicationComponent.kt
package net.squanchy.injection

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import net.squanchy.analytics.Analytics
import net.squanchy.analytics.AnalyticsModule
import net.squanchy.remoteconfig.FeatureFlags
import net.squanchy.remoteconfig.RemoteConfig
import net.squanchy.remoteconfig.RemoteConfigModule
import net.squanchy.schedule.tracksfilter.TracksFilter
import net.squanchy.schedule.tracksfilter.TracksFilterModule
import net.squanchy.search.algolia.AlgoliaModule
import net.squanchy.search.algolia.AlgoliaSearchEngine
import net.squanchy.service.firebase.FirestoreDbService
import net.squanchy.service.firebase.injection.FirestoreModule
import net.squanchy.service.repository.AuthService
import net.squanchy.service.repository.EventRepository
import net.squanchy.service.repository.SpeakerRepository
import net.squanchy.service.repository.TracksRepository
import net.squanchy.service.repository.injection.RepositoryModule
import net.squanchy.support.checksum.ChecksumModule
import net.squanchy.support.injection.CurrentTimeModule
import net.squanchy.support.system.CurrentTime

fun createApplicationComponent(application: Application): ApplicationComponent {
    return DaggerApplicationComponent.builder()
        .application(application)
        .build()
}

@ApplicationLifecycle
@Component(
    modules = [
        AlgoliaModule::class,
        FirestoreModule::class,
        ChecksumModule::class,
        RepositoryModule::class,
        AnalyticsModule::class,
        RemoteConfigModule::class,
        CurrentTimeModule::class,
        TracksFilterModule::class
    ]
)
interface ApplicationComponent {

    fun firestoreDbService(): FirestoreDbService

    fun firebaseAuthService(): AuthService

    fun eventRepository(): EventRepository

    fun speakerRepository(): SpeakerRepository

    fun tracksRepository(): TracksRepository

    fun tracksFilter(): TracksFilter

    fun analytics(): Analytics

    fun remoteConfig(): RemoteConfig

    fun featureFlags(): FeatureFlags

    fun application(): Application

    fun algoliaSearchEngine(): AlgoliaSearchEngine

    fun currentTime(): CurrentTime

    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(app: Application): Builder

        fun build(): ApplicationComponent
    }
}
